### PR TITLE
[663] Allow unit rename on promote

### PR DESCRIPTION
**Pull request by davidhoward**
* Add nullable "instanceName" field to MapUnit to capture name for individual units
* Add "displayName" property to MapUnit which takes instanceName if set, name
  otherwise
* Add dialog to PromotionPicker to allow setting instanceName of unit, if not
  already set

### CODE REVIEW

File: core/src/com/unciv/logic/map/MapUnit.kt [hunk 1]
```diff
@@ -57,7 +61,22 @@ class MapUnit {
...
+
+    /**
+     *  Name of this individual unit, usually resulting from promotion
+     */
+    var instanceName: String? = null
+
+    /**
+     * Name which should be displayed in UI
+     */
+    val displayName: String get() = instanceName?: name
```
Comment (yairm210) [inline core/src/com/unciv/logic/map/MapUnit.kt:78]: As part of the coding standard in this project, information which is not saved in the game state should be either transient or converted to a function - please change this to a function.

File: core/src/com/unciv/logic/battle/Battle.kt [hunk 2]
```diff
@@ -429,7 +429,7 @@ object Battle {
             interceptor.attacksThisTurn++
 
             val attackerName = attacker.getName()
-            val interceptorName = interceptor.name
+            val interceptorName = interceptor.displayName
```
Comment (yairm210) [inline core/src/com/unciv/logic/battle/Battle.kt:432]: The interceptor name should be relevant only to the civ who the interceptor belongs to

File: core/src/com/unciv/logic/battle/MapUnitCombatant.kt [hunk 3]
```diff
@@ -10,7 +10,7 @@ class MapUnitCombatant(val unit: MapUnit) : ICombatant {
     override fun getMaxHealth() = 100
     override fun getCivInfo(): CivilizationInfo = unit.civInfo
     override fun getTile(): TileInfo = unit.getTile()
-    override fun getName(): String = unit.name
+    override fun getName(): String = unit.displayName
```
Comment (yairm210) [inline core/src/com/unciv/logic/battle/MapUnitCombatant.kt:13]: This too is context-specific, so I wouldn't change this.

File: core/src/com/unciv/logic/map/MapUnit.kt [hunk 4]
```diff
@@ -74,11 +93,20 @@ class MapUnit {
         private const val ANCIENT_RUIN_MAP_REVEAL_CHANCE = 0.8f
     }
 
+    // GDX serialization needs a no-arg constructor to be defined, if any constructor is.
+    constructor() {}
+
+    constructor(baseUnit_ : BaseUnit, civInfo_: CivilizationInfo) {
```
Comment (yairm210) [inline core/src/com/unciv/logic/map/MapUnit.kt:99]: In what situation is this actually necessary? It seems that you've changed a lot of code to fit this patters, but it's unclear why the pattern is needed at all. In any case this change (and all related changes) should be part of a separate PR.

File: core/src/com/unciv/ui/pickerscreens/PromotionPickerScreen.kt [hunk 5]
```diff
@@ -52,6 +53,16 @@ class PromotionPickerScreen(val unit: MapUnit) : PickerScreen() {
                     || unit.promotions.promotions.contains(it.name) }
         val unitAvailablePromotions = unit.promotions.getAvailablePromotions()
 
+        if(canPromoteNow && unit.instanceName == null) {
+            val renameButton = Button(skin)
+            renameButton.add("Choose name for [${unit.name}]".tr())
```
Comment (yairm210) [inline core/src/com/unciv/ui/pickerscreens/PromotionPickerScreen.kt:58]: `val renameButton = "Choose name for [${unit.name}]".toTextButton()

File: core/src/com/unciv/ui/pickerscreens/RenameUnitPopup.kt [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+import com.badlogic.gdx.scenes.scene2d.ui.TextField
+import com.unciv.logic.map.MapUnit
+import com.unciv.models.translations.tr
+import com.unciv.ui.utils.*
+
+class RenameUnitPopup(unit: MapUnit, screen: CameraStageBaseScreen): Popup(screen) {
+
+    init {
+        add("Choose name for [${unit.baseUnit.name}]".tr().toLabel())
```
Comment (yairm210) [inline core/src/com/unciv/ui/pickerscreens/RenameUnitPopup.kt:11]: addGoodSizedLabel(...).row()

**Comment by yairm210** [comment]
This is a good start!

A. There's a hidden problem here - if you call your Warrior, say, "Archer", then all your enemy's notifications will be "Your X was attacked by an enemy Archer" which is worse than unhelpful.
There are 2 way to solve this:
1. Add the unit type in brackets - "Archer (Warrior)"
2. Only display the unit name to the owning civ - this is what I put in the comments, but then I saw how many code changes this would require and I think option 1 is better

B. Unit name should be limited in size (10 chars?) or the UI will break in multiple places

C. The constructor change seems unnecessary, and is definitely confusing - if it's needed should go in another PR

File: core/src/com/unciv/logic/map/MapUnit.kt (hunk 4 above)
Comment (davidhoward) [reply core/src/com/unciv/logic/map/MapUnit.kt:99]: This is kind-of a holdover from an early change I tried to remove `name` altogether, before I came to understand that it needed to exist and be publically settable in order for deserialization to work. 

I still have some concern about the possibility of  `baseUnit` and `name`  or `civInfo` and `owner` getting out of sync, but you're right that's an issue for another PR.

File: core/src/com/unciv/ui/pickerscreens/RenameUnitPopup.kt [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+        val nameField = TextField("", skin)
+        // Disallowed special characters determined with US keyboard and a certain binary
+        // copyrighted by Firaxis Games
+        val illegal_chars = "%*[]\"\\|<>/?"
+        nameField.textFieldFilter = TextField.TextFieldFilter { _, char -> !(char in illegal_chars)}
+        // Max name length decided arbitrarily
+        nameField.maxLength = 10
+        add(nameField)
+        row()
+        add("Ok".toTextButton().onClick {
```
Comment (yairm210) [inline core/src/com/unciv/ui/pickerscreens/RenameUnitPopup.kt:22]: Change to "OK", which is already in the translation file - translations are case-sensitive

File: core/src/com/unciv/ui/pickerscreens/RenameUnitPopup.kt [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+            if (nameField.text != "") {
+                unit.instanceName = nameField.text
+            }
+            screen.game.setScreen(PromotionPickerScreen(unit))
+        })
+        add("Cancel".toTextButton().onClick { close() })
```
Comment (yairm210) [inline core/src/com/unciv/ui/pickerscreens/RenameUnitPopup.kt:28]: Either change to "Close", which is already in the translation file, or add "Cancel" as a translation line in template.properties

**Comment by yairm210** [comment]
Looks good, finish up with the translations and we'll merge this!

I do have some nitpicky comments about code syntax, like !(char in illegal_chars) -> char !in illegal_chars and add().row() in a single line, but there's nothing real to prevent us merging this :)

**Comment by davidhoward** [comment]
Sounds good. In the future I'm open to nitpick comments, this is my first kotlin project so style guidance is appreciated!

Also, I've been amending/squashing my commits as I respond to review comments. Is this the preferred method for the project? I didn't see anything about it in the guide

**Comment by yairm210** [comment]
It's fine whatever you do, since we squash the final merge into a single commit anyway :)
